src/migration/migration.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <filesystem>
+#include <fstream>
+
+#include "migration/migration.hpp"
+
+using namespace Poco::Data::Keywords;
+
+Migration::Migration(Poco::Data::Session& session, const std::string& migrationDir)
+    : session_(session)
+    , migrationDir_(migrationDir)
+{
+    CreateVersionTableIfNeeded();
+}
+
+void Migration::MigrateToVersion(int targetVersion)
+{
+    int currentVersion = GetCurrentVersion();
+
+    if (currentVersion == targetVersion) {
+        return;
+    }
+
+    if (currentVersion < targetVersion) {
+        UpgradeDatabase(targetVersion, currentVersion);
+
+        return;
+    }
+
+    DowngradeDatabase(targetVersion, currentVersion);
+}
+
+void Migration::ApplyMigration(const std::string& migrationScript)
+{
+    std::string   path = std::filesystem::path(migrationDir_) / migrationScript;
+    std::ifstream file(path);
+
+    if (!file.is_open()) {
+        throw std::runtime_error("Failed to open migration script: " + migrationScript);
+    }
+
+    std::string script((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+
+    session_ << script, now;
+}
+
+void Migration::UpgradeDatabase(int targetVersion, int currentVersion)
+{
+    for (int i = currentVersion + 1; i <= targetVersion; i++) {
+        ApplyMigration(std::to_string(i) + "_update.up.sql");
+        UpdateVersion(i);
+    }
+}
+
+void Migration::DowngradeDatabase(int targetVersion, int currentVersion)
+{
+    for (int i = currentVersion; i > targetVersion; i--) {
+        ApplyMigration(std::to_string(i) + "_update.down.sql");
+        UpdateVersion(i - 1);
+    }
+}
+
+void Migration::CreateVersionTableIfNeeded()
+{
+    session_ << "CREATE TABLE IF NOT EXISTS SchemaVersion (version INTEGER);", now;
+    session_ << "INSERT INTO SchemaVersion (version) SELECT 0 WHERE NOT EXISTS "
+                "(SELECT 1 FROM SchemaVersion)",
+        now;
+}
+
+void Migration::UpdateVersion(int version)
+{
+    session_ << "UPDATE SchemaVersion SET version = ?;", use(version), now;
+}
+
+int Migration::GetCurrentVersion()
+{
+    int                   version = 0;
+    Poco::Data::Statement select(session_);
+
+    select << "SELECT version FROM SchemaVersion LIMIT 1;", into(version), now;
+
+    return version;
+}
```

Comment:
Please add empty line at the end of the file

Reply:
Done